File: LeetCode/Python/Hard/2402-Meeting-Rooms-III/solution.py
import heapq

class Solution:
    def mostBooked(self, n: int, meetings: List[List[int]]) -> int:
        result=0

        meetings.sort(key=lambda meeting: meeting[0])

        meetings_held_per_room=[0 for _ in range(0, n)]

        available_meeting_rooms=list()

        rooms_in_use=list()

        for meeting_room in range(0, n):
            heapq.heappush(available_meeting_rooms, meeting_room)

        most_meetings_held=0

        for meeting in meetings:
            current_meeting_start=meeting[0]

            current_meeting_end=meeting[1]

            while rooms_in_use and rooms_in_use[0][0] <= current_meeting_start:
                (ending_time, room)=heapq.heappop(rooms_in_use)

                heapq.heappush(available_meeting_rooms, room)

            if available_meeting_rooms:
                room=heapq.heappop(available_meeting_rooms)

                meetings_held_per_room[room]+=1

                heapq.heappush(rooms_in_use, (current_meeting_end, room))

                most_meetings_held=max(most_meetings_held, meetings_held_per_room[room])
            else:
                (ending_time, room)=heapq.heappop(rooms_in_use)

                meetings_held_per_room[room]+=1

                current_meeting_duration=current_meeting_end - current_meeting_start

                delayed_meeting_ending_time=ending_time + current_meeting_duration

                heapq.heappush(rooms_in_use, (delayed_meeting_ending_time, room))

                most_meetings_held=max(most_meetings_held, meetings_held_per_room[room])

        for room in range(0, n):
            if meetings_held_per_room[room]==most_meetings_held:
                result=room

                break

        return result

File: LeetCode/Rust/Hard/2402-Meeting-Rooms-III/solution.rs
use std::collections::{BinaryHeap};
use std::cmp;

impl Solution
{
    pub fn most_booked(n: i32, mut meetings: Vec<Vec<i32>>) -> i32
    {
        let mut result: i32=0;

        meetings.sort();

        let mut meetings_held_per_room: Vec<i32>=vec![0; n as usize];

        let mut available_meeting_rooms: BinaryHeap<std::cmp::Reverse<i32>>=BinaryHeap::new();

        let mut rooms_in_use: BinaryHeap<std::cmp::Reverse<(i64, i32)>>=BinaryHeap::new();

        for meeting_room in 0..n
        {
            available_meeting_rooms.push(std::cmp::Reverse(meeting_room));
        }

        let mut most_meetings_held: i32=0;

        for meeting in meetings.iter()
        {
            let current_meeting_start: i64=meeting[0] as i64;

            let current_meeting_end: i64=meeting[1] as i64;

            while let Some(std::cmp::Reverse((ending_time, room)))=rooms_in_use.peek()
            {
                if *ending_time <= current_meeting_start
                {
                    if let Some(std::cmp::Reverse((ending_time, room)))=rooms_in_use.pop()
                    {
                        available_meeting_rooms.push(std::cmp::Reverse(room));
                    }
                }
                else
                {
                    break;
                }
            }

            if let Some(std::cmp::Reverse(room))=available_meeting_rooms.pop()
            {
                meetings_held_per_room[room as usize]+=1;

                rooms_in_use.push(std::cmp::Reverse((current_meeting_end, room)));

                most_meetings_held=std::cmp::max(most_meetings_held, meetings_held_per_room[room as usize]);
            }
            else
            {
                if let Some(std::cmp::Reverse((ending_time, room)))=rooms_in_use.pop()
                {
                    meetings_held_per_room[room as usize]+=1;

                    let current_meeting_duration: i64=current_meeting_end - current_meeting_start;

                    let delayed_meeting_ending_time: i64=ending_time + current_meeting_duration;

                    rooms_in_use.push(std::cmp::Reverse((delayed_meeting_ending_time, room)));

                    most_meetings_held=std::cmp::max(most_meetings_held, meetings_held_per_room[room as usize]);
                }
            }
        }

        for room in 0..n
        {
            if meetings_held_per_room[room as usize]==most_meetings_held
            {
                result=room;

                break;
            }
        }

        result
    }
}